src/tests/test_raft_repl_dev.cpp

```diff
@@ -331,45 +331,132 @@ TEST_F(RaftReplDevTest, All_Append_Fetch_Remote_Data) {
 
     g_helper->sync_for_verify_start();
 
-    // TODO: seems with filip and fetch remote, the data size is not correct;
     LOGINFO("Validate all data written so far by reading them");
     this->validate_all_data();
+
     g_helper->sync_for_cleanup_start();
 }
 
-TEST_F(RaftReplDevTest, All_ReplService) {
+// do some io before restart;
+TEST_F(RaftReplDevTest, All_restart_one_follower_inc_resync) {
     LOGINFO("Homestore replica={} setup completed", g_helper->replica_num());
     g_helper->sync_for_test_start();
-    auto repl_dev = dynamic_pointer_cast< RaftReplDev >(pick_one_db().repl_dev());
-    auto group_id = repl_dev->group_id();
-    auto my_id_str = repl_dev->my_replica_id_str();
 
-    auto leader = repl_dev->get_leader_id();
-    ASSERT_TRUE(leader != replica_id_t())
-        << "Error getting leader id for group_id=" << boost::uuids::to_string(group_id).c_str();
-    auto leader_str = boost::uuids::to_string(leader);
-    LOGINFO("Got raft leader {} for group {}", leader_str, group_id);
+    // step-0: do some IO before restart one member;
+    uint64_t exp_entries = 20;
+    if (g_helper->replica_num() == 0) {
+        g_helper->runner().set_num_tasks(20);
+        auto block_size = SISL_OPTIONS["block_size"].as< uint32_t >();
+        LOGINFO("Run on worker threads to schedule append on repldev for {} Bytes.", block_size);
+        g_helper->runner().set_task([this, block_size]() {
+            this->generate_writes(block_size /* data_size */, block_size /* max_size_per_iov */);
+        });
+        g_helper->runner().execute().get();
+    }
 
+    // step-1: wait for all writes to be completed
+    this->wait_for_all_writes(exp_entries);
+
+    // step-2: restart one non-leader replica
+    if (g_helper->replica_num() == 1) {
+        LOGINFO("Restart homestore: replica_num = 1");
+        g_helper->restart();
+        g_helper->sync_for_test_start();
+    }
+
+    exp_entries += SISL_OPTIONS["num_io"].as< uint64_t >();
+    // step-3: on leader, wait for a while for replica-1 to finish shutdown so that it can be removed from raft-groups
+    // and following I/O issued by leader won't be pushed to relica-1;
     if (g_helper->replica_num() == 0) {
-        ASSERT_TRUE(leader_str == my_id_str)
-            << "Leader id " << leader_str.c_str() << " should  equals to my ID " << my_id_str.c_str();
-    } else {
-        ASSERT_TRUE(leader_str != my_id_str) << "I am a follower, Leader id " << leader_str.c_str()
-                                             << " should not equals to my ID " << my_id_str.c_str();
+        LOGINFO("Wait for grpc connection to replica-1 to expire and removed from raft-groups.");
+        std::this_thread::sleep_for(std::chrono::seconds{5});
```

Comment:
I feel instead of 5 seconds sleep, we can query the repl_dev to get the number of alive members as 2 and then proceed?

Reply:
the issue I dealt with is replica-1 is already gone, but shutdown takes longer time (take cp, etc.) and if we start I/O immediately after repilca-1 is gone, the 300 I/O finishes too fast before replica-1 is back. I want to simulate resync while I/Os are still on-going.